dev/TabView/TabViewItem.cpp

```diff
@@ -106,15 +129,90 @@ winrt::AutomationPeer TabViewItem::OnCreateAutomationPeer()
     return winrt::make<TabViewItemAutomationPeer>(*this);
 }
 
+void TabViewItem::OnCloseButtonOverlayModeChanged(winrt::TabViewCloseButtonOverlayMode const& mode)
+{
+    m_closeButtonOverlayMode = mode;
+    if (mode == winrt::TabViewCloseButtonOverlayMode::OnHover && !IsSelected())
+    {
+        UpdateCloseButtonVisibility(winrt::Visibility::Collapsed);
+    }
+    if (mode == winrt::TabViewCloseButtonOverlayMode::Persistent || mode == winrt::TabViewCloseButtonOverlayMode::Auto)
+    {
+        UpdateCloseButtonVisibility(winrt::Visibility::Visible);
+    }
+}
+
+void TabViewItem::OnTabViewWidthModeChanged(winrt::TabViewWidthMode const& mode)
+{
+    m_tabViewWidthode = mode;
+
+    // When switching to compact, all items hide the labels and only show their icon
+    // This switch needs to be done here!
+    if (mode == winrt::TabViewWidthMode::Compact)
+    {
+        if (!IsSelected())
+        {
+            winrt::VisualStateManager::GoToState(*this, L"Compact", false);
+        }
+    }
+    else
+    {
+        winrt::VisualStateManager::GoToState(*this, L"StandardWidth",false);
+    }
+}
+
 
 void TabViewItem::UpdateCloseButton()
 {
+    if (!IsClosable())
+    {
+        UpdateCloseButtonVisibility(winrt::Visibility::Collapsed);
+    }
+    else
+    {
+        switch (m_closeButtonOverlayMode)
+        {
+            case winrt::TabViewCloseButtonOverlayMode::Persistent:
+            {
+                // In case of "Persistent" always show button
+                UpdateCloseButtonVisibility(winrt::Visibility::Visible);
+                break;
+            }
+            case winrt::TabViewCloseButtonOverlayMode::Auto:
+            {
+                // In case of "Auto" always show button
+                UpdateCloseButtonVisibility(winrt::Visibility::Visible);
+                break;
+            }
+            case winrt::TabViewCloseButtonOverlayMode::OnHover:
+            {
+                // If we only want to show the button on hover, we also show it when we are selected, otherwise hide it
+                if (IsSelected())
+                {
+                    UpdateCloseButtonVisibility(winrt::Visibility::Visible);
+                }
+                else
+                {
+                    UpdateCloseButtonVisibility(winrt::Visibility::Collapsed);
+                }
+                break;
+            }
+        }
+    }
     if (auto && closeButton = m_closeButton.get())
     {
         closeButton.Visibility(IsClosable() ? winrt::Visibility::Visible : winrt::Visibility::Collapsed);
     }
 }
 
+void TabViewItem::UpdateCloseButtonVisibility(winrt::Visibility const& newVisibility)
```

Comment:
Instead of this method, we should handle it as a visual state.

Reply:
Is there any benefit to handling this inside a VisualState? After all there is quite a lot of logic involved in determining the correct state, and I am not sure how easy transferring that into VisualStates would be.

Reply:
I would leave the logic in code for sure, just instead of calling UpdateCloseButtonVisibility(), call GoToState() with either "CloseButtonVisible" or "CloseButtonCollapsed" something like that.

Reply:
I have updated the code and switch to the VisualStates now.